src/kryptos/k4/ene_routes.py

```
@@ -0,0 +1,127 @@
+    """Discrete rational dcol/drow approximation for a compass bearing.
+
+    North (0 deg) advances straight down the grid with no column drift;
+    ENE (67.5 deg) yields ~2.414 (matching the existing ``tan(67.5)``
+    constant used elsewhere), NE (45 deg) yields 1.
+    """
+    deg = _bearing_degrees(bearing)
+    theta = math.radians(deg)
+    drow = math.cos(theta)
+    dcol = math.sin(theta)
+    if abs(drow) < 1e-9:
+        raise ValueError(f"bearing {deg} has no row progression (due E/W)")
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)